### DISCUSSION TITLE
How to get bytes instead of str inside span?

### DISCUSSION THREAD

**Discussion by ivanstepanovftw** [pymupdf/PyMuPDF#2834] [Q&A]
I need to get raw `bytes` that are inside span. `span["text"]` is of type `str`. `span["chars"][N]["c"]` is of type `str` too. What should I do to get raw bytes?
```py
INVALID_UNICODE = chr(0xFFFD)  # the "Invalid Unicode" character
doc = fitz.fitz.Document("pdfs/arabic.pdf")

for page in doc:
    blocks = fitz.utils.get_text(
        page=page,
        option="dict",
        flags=0
    )["blocks"]
    for block in blocks:
        for line in block["lines"]:
            for span in line["spans"]:
                text = span["text"]
                if INVALID_UNICODE in text:
                    # parse error
```

PDF: http://tug.ctan.org/macros/latex/exptl/mem/arabic.pdf

---

This script is from https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/OCR/tesseract2.py, and the idea I am wanted to do is to extract bytes from span character and try to search for relevant glyph in the current font (that is of `.cff` file type), then render the glyph for OCR.
FontForge screenshot:
![image](https://github.com/pymupdf/PyMuPDF/assets/8203898/a42616d5-ed18-40e7-9304-1e08e2274d7a)

**Comment by JorjMcKie** [comment]
This no issue, but a **_Discussions_** post. Transferring ...

**Reply by JorjMcKie** [reply]
Using `get_text("rawdict",...)` delivers the exact same structure as "dict". You could use "rawdict" always. Although the key `span["text"]` is missing, that text can easily be generated by one statement `text = "".join([c["c"] for c in span["chars"]])`.

**Reply by ivanstepanovftw** [reply]
I am asking about `bytes`, not `str`. While the library always using `str`, I am getting `\xEF\xBF\xBD`, that is the UTF-8 encoding for the unicode character `U+FFFD` (Replacement character). 

Does PDF contain this string? 
Or it is a missing PyMuPDF feature for extracting raw bytes?
Or is it a bug? I see a relevant issue https://github.com/pymupdf/PyMuPDF/issues/365

**Reply by ivanstepanovftw** [reply]
The PDF I am referring to does not contain "Replacement character".

**Reply by JorjMcKie** [reply]
Ah, ok. The replacement character 0xfffd is generated. This happens in cases when for example surrogate Unicodes have been used illegally, or glyph numbers have been used for which the font has no backtranslation to unicode numbers.
When this happens, there is no way to access the original bytes in the PDF.

**Reply by ivanstepanovftw** [reply]
I see that PDF indeed contains raw bytes in the complex script.
![image](https://github.com/pymupdf/PyMuPDF/assets/8203898/5458a056-ebd4-4baf-b6db-3cfe3a3dcce5)

And it is not replacement character:
![image](https://github.com/pymupdf/PyMuPDF/assets/8203898/b5e28755-8f7b-4df4-98e3-afd7dc1bc81d)


Added issue https://github.com/pymupdf/PyMuPDF/issues/2835

**Reply by ivanstepanovftw** [reply]
Even Google Chrome can do this, but as a text, and, possibly, not raw bytes.
![image](https://github.com/pymupdf/PyMuPDF/assets/8203898/2560e539-cf72-4d77-b6e0-30e7c9116cb1)
